storage/rocksdb/rdb_iterator.cc

```diff
@@ -0,0 +1,974 @@
+  /*
+    We are looking for the first record such that
+
+      index_tuple $LT lookup_tuple
+
+    with HA_READ_BEFORE_KEY, $LT = '<',
+    with HA_READ_PREFIX_LAST_OR_PREV, $LT = '<='
+    with HA_READ_PREFIX_LAST, $LT = '=='
+
+    Symmetry with read_after_key is possible if rocksdb supported prefix seeks.
+  */
+  rocksdb_smart_seek(!m_kd->m_is_reverse_cf, m_scan_it, key_slice);
+
+  while (is_valid_iterator(m_scan_it)) {
+    if (thd_killed(m_thd)) {
+      return HA_ERR_QUERY_INTERRUPTED;
+    }
+    /*
+      We are using full key and we've hit an exact match.
+      */
+    if ((full_key_match &&
+         m_kd->value_matches_prefix(m_scan_it->key(), key_slice))) {
+      rocksdb_smart_next(!m_kd->m_is_reverse_cf, m_scan_it);
+      continue;
+    }
+
+    return HA_EXIT_SUCCESS;
+  }
+
+  return iter_status_to_retval(m_scan_it, m_kd, HA_ERR_END_OF_FILE);
+}
+
+int Rdb_iterator_base::read_after_key(const rocksdb::Slice &key_slice) {
+  /*
+    We are looking for the first record such that
+
+    index_tuple $GT lookup_tuple
+
+    with HA_READ_AFTER_KEY, $GT = '>',
+    with HA_READ_KEY_OR_NEXT, $GT = '>='
+    with HA_READ_KEY_EXACT, $GT = '=='
+  */
+  rocksdb_smart_seek(m_kd->m_is_reverse_cf, m_scan_it, key_slice);
+
+  return is_valid_iterator(m_scan_it) ?
+             HA_EXIT_SUCCESS : 
+             iter_status_to_retval(m_scan_it, m_kd, HA_ERR_END_OF_FILE);
+}
+
+void Rdb_iterator_base::release_scan_iterator() {
+  delete m_scan_it;
+  m_scan_it = nullptr;
+  m_use_locking_iter = false;
+
+  if (m_scan_it_snapshot) {
+    auto rdb = rdb_get_rocksdb_db();
+    rdb->ReleaseSnapshot(m_scan_it_snapshot);
+    m_scan_it_snapshot = nullptr;
+  }
+}
+
+void Rdb_iterator_base::setup_scan_iterator(const rocksdb::Slice *const slice,
+                                            const uint eq_cond_len,
+                                            bool read_current) {
+  DBUG_ASSERT(slice->size() >= eq_cond_len);
+
+  bool skip_bloom = true;
+
+  const rocksdb::Slice eq_cond(slice->data(), eq_cond_len);
+
+  // The size of m_scan_it_lower_bound (and upper) is technically
+  // max_packed_sk_len as calculated in ha_rocksdb::alloc_key_buffers.  Rather
+  // than recalculating that number, we pass in the max of eq_cond_len and
+  // Rdb_key_def::INDEX_NUMBER_SIZE which is guaranteed to be smaller than
+  // max_packed_sk_len, hence ensuring no buffer overrun.
+  //
+  // See setup_iterator_bounds on how the bound_len parameter is
+  // used.
+  if (ha_rocksdb::check_bloom_and_set_bounds(
+          m_thd, *m_kd, eq_cond,
+          std::max(eq_cond_len, (uint)Rdb_key_def::INDEX_ID_SIZE),
+          m_scan_it_lower_bound, m_scan_it_upper_bound,
+          &m_scan_it_lower_bound_slice, &m_scan_it_upper_bound_slice)) {
+    skip_bloom = false;
+  }
+
+  // Save the value of m_use_locking_iter because release_scan_iterator()
+  // will set it to false.
+  bool use_locking_iter= m_use_locking_iter;
```

Comment:
It doesn't look like m_use_locking_iter is set back to use_locking_iter after release_scan_iterator() is called. 

In index_read_intern(), for read-committed, we could call setup_scan_iterator and release_iterator multiple times. If we ever did that for repeatable-read, then it seems m_use_locking_iter, then while the first couple of times through the loop might grab a locking iterator, the third and later times through the loop would not allocate a locking iterator.

Reply:
Addressed